packages/ember-data/lib/serializers/rest-serializer.js

```diff
@@ -689,6 +689,49 @@ var RESTSerializer = JSONSerializer.extend({
     } else {
       json[key + "Type"] = Ember.String.camelize(belongsTo.modelName);
     }
+  },
+
+  /**
+    You can use this method to customize how a polymorphic relationship should
+    be extracted.
+
+    @method extractPolymorphicRelationship
+    @param {Object} relationshipModelName
+    @param {Object} relationshipHash
+    @param {Object} relationshipOptions
+    @return {Object}
+   */
+  extractPolymorphicRelationship: function(relationshipModelName, relationshipHash, relationshipOptions) {
+    var { relationshipKey, resourceHash, relationshipMeta } = relationshipOptions;
+
+    // A polymorphic belongsTo relationship can be present in the payload
+    // either in the form where the `id` and the `type` are given:
+    //
+    //   {
+    //     message: { id: 1, type: 'post' }
+    //   }
+    //
+    // or by the `id` and a `<relationship>Type` attribute:
+    //
+    //   {
+    //     message: 1,
+    //     messageType: 'post'
+    //   }
+    //
+    // The next code checks if the latter case is present and returns the
+    // corresponding JSON-API representation. The former case is handled within
+    // the super class (JSONSerializer).
+    var isPolymorphic = relationshipMeta.options.polymorphic;
+    var typeProperty = `${relationshipKey}Type`;
```

Comment:
I think this should probably go through `keyForAttribute`. I think we had a bug related to this previously. It would also make the ActiveModelSerializer work out of the box... I think.